### PR TITLE
Add `OnceCell`

### PR DESCRIPTION
**Pull request by SabrinaJewson**

### CODE REVIEW

File: src/once_cell.rs [hunk 1]
```diff
@@ -0,0 +1,431 @@
...
+            State::Initialized { value } => Some(value),
+        }
+    }
+}
+
+impl<T: Debug> Debug for OnceCell<T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let mut s = f.debug_struct("OnceCell");
+        if let Some(inner) = self.get() {
+            s.field("value", inner);
```
Comment (udoprog) [inline src/once_cell.rs:263]: Might as well debug the `state` field as-is. We can then have `#[derive(Debug)]`.

File: src/wait_list.rs [hunk 2]
```diff
@@ -313,20 +313,22 @@ impl<'wait_list, I, O> Borrowed<'wait_list, I, O> {
...
-            Some(head) => head,
-            None => return,
-        };
+    ///
+    /// Returns ownership of that waiter's input value.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error when there are no wakers in the list.
+    pub fn wake_one(&mut self, output: O) -> Result<I, ()> {
```
Comment (udoprog) [inline src/wait_list.rs:322]: Might as well name error type, even if it's a `#[non_exhaustive]` unit struct. Helps with readability.

File: src/once_cell.rs [hunk 3]
```diff
@@ -0,0 +1,431 @@
...
+    pub const fn new() -> Self {
+        Self {
+            state: UnsafeCell::new(State::Uninit),
+            waiters: WaitList::new(),
+        }
+    }
+
+    /// Get a shared reference to the inner value, returning `None` if it has not been set yet or is
+    /// in the process of being set.
+    #[must_use]
```
[the remark this answers is not in the document]
Comment (SabrinaJewson) [reply src/once_cell.rs:50]: This function doesn't have any side effects so I don't think it's correct to ever have `once_cell.get();` on its own — it makes sense to warn that I think.

File: src/wait_list.rs (hunk 2 above)
Comment (SabrinaJewson) [reply src/wait_list.rs:322]: An alternative is passing back `O` to the caller. What do you think about `Result<I, WakeOneError<O>>` versus `Result<I, O>`?

Comment (SabrinaJewson) [reply src/wait_list.rs:322]: @udoprog Any thoughts on this?

Comment (udoprog) [reply src/wait_list.rs:322]: I'd always prefer a named error. It gives the ability to format the message when printed or unwrapped. What data it returns is up to what's needed by users.

**Pull request merged**